controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,23 +194,23 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
 
 // MatchesKCPConfiguration returns a filter to find all machines that matches with KCP config and do not require any rollout.
 // Kubernetes version, infrastructure template, and KubeadmConfig field need to be equivalent.
-func MatchesKCPConfiguration(ctx context.Context, c client.Client, kcp *controlplanev1.KubeadmControlPlane) func(machine *clusterv1.Machine) bool {
+func MatchesKCPConfiguration(infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane) func(machine *clusterv1.Machine) bool {
 	return And(
 		MatchesKubernetesVersion(kcp.Spec.Version),
-		MatchesKubeadmBootstrapConfig(ctx, c, kcp),
-		MatchesTemplateClonedFrom(ctx, c, kcp),
+		MatchesKubeadmBootstrapConfig(machineConfigs, kcp),
+		MatchesTemplateClonedFrom(infraConfigs, kcp),
 	)
 }
 
 // MatchesTemplateClonedFrom returns a filter to find all machines that match a given KCP infra template.
-func MatchesTemplateClonedFrom(ctx context.Context, c client.Client, kcp *controlplanev1.KubeadmControlPlane) Func {
+func MatchesTemplateClonedFrom(infraConfigs map[string]*unstructured.Unstructured, kcp *controlplanev1.KubeadmControlPlane) Func {
```

Comment:
It feels odd to pass in a map of infrastructure resources here, since it generates a function that takes a single Machine.

Reply:
Not sure what you mean, this generates a partially applied function that needs to be able to handle any machine, so we pass in/memoize the map of all infrastructure resources for the given cluster so that we can do a lookup on each call.